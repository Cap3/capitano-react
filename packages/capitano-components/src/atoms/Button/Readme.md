```js
const { Grid, EllipsisText } = require("../index");

<Grid columnCount={4} gap={20}>
  <Button>Primary</Button>
  <Button type="secondary">Secondary</Button>
  <Button type="success">Success</Button>
  <Button type="danger">Danger</Button>
  <Button outline>Primary outline</Button>
  <Button outline type="secondary">
    Secondary outline
  </Button>
  <Button outline type="success">
    Success outline
  </Button>
  <Button outline type="danger">
    Danger outline
  </Button>
  <Button disabled>Primary disabled</Button>
  <Button disabled type="secondary">
    Secondary disabled
  </Button>
  <Button disabled outline type="success">
    Success/outline disabled
  </Button>
  <Button disabled outline type="danger">
    Danger/outline disabled
  </Button>
  <Button fixed>
    <EllipsisText>Primary fixed</EllipsisText>
  </Button>
  <Button fixed type="secondary">
    <EllipsisText>Secondary fixed</EllipsisText>
  </Button>
  <Button fixed type="success">
    <EllipsisText>Success fixed</EllipsisText>
  </Button>
  <Button fixed type="danger">
    <EllipsisText>Danger fixed</EllipsisText>
  </Button>
</Grid>;
```
