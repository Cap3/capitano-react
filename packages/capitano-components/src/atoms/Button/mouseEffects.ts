import { CapitanoTheme, isColorEffect } from '@cap3/capitano-theme';
import { ColorInput, TinyColor } from '@ctrl/tinycolor';
import { CSSProperties } from 'react';
import { topShadow, bottomShadow } from '../basics/shadows';

type EffectColorArgs = {
  baseColor: ColorInput;
  swapped: boolean;
  outline: boolean;
  disabled?: boolean;
  theme: CapitanoTheme;
};

const effectColorStyle = (
  disabled: boolean,
  swapped: boolean,
  color: string,
  effect: 'hover' | 'focus' | 'active' | 'focus-within' | 'visited',
  custom?: CSSProperties,
) => {
  return {
    [`&:${effect}`]: {
      backgroundColor: disabled || swapped ? undefined : color,
      color: disabled || !swapped ? undefined : color,
      borderColor: disabled || !swapped ? undefined : color,
      ...custom,
    },
  };
};

export const focusEffect = ({
  disabled = false,
  theme,
  baseColor,
  swapped,
  outline,
}: EffectColorArgs) => {
  const base = new TinyColor(baseColor);

  let background: string;

  if (isColorEffect(theme.statusEffects.focus)) {
    background = base[theme.statusEffects.focus.effect](
      theme.statusEffects.focus.factor,
    ).toRgbString();
  } else {
    background = base
      .mix(theme.statusEffects.focus.second, theme.statusEffects.focus.factor)
      .toRgbString();
  }

  return effectColorStyle(
    disabled,
    swapped,
    background,
    'focus',
    outline
      ? {
          backgroundColor: 'transparent',
        }
      : undefined,
  );
};

export const activeEffect = ({
  disabled = false,
  theme,
  baseColor,
  swapped,
  outline,
}: EffectColorArgs) => {
  const base = new TinyColor(baseColor);
  let background: string;

  if (isColorEffect(theme.statusEffects.pressed)) {
    background = base[theme.statusEffects.pressed.effect](
      theme.statusEffects.pressed.factor,
    ).toRgbString();
  } else {
    background = base
      .mix(theme.statusEffects.pressed.second, theme.statusEffects.pressed.factor)
      .toRgbString();
  }

  return effectColorStyle(
    disabled,
    swapped,
    background,
    'active',
    outline
      ? {
          backgroundColor: 'transparent',
        }
      : undefined,
  );
};

export const hoverEffect = ({
  disabled = false,
  theme,
  baseColor,
  swapped,
  outline,
}: EffectColorArgs) => {
  const base = new TinyColor(baseColor);
  let background: string;

  if (isColorEffect(theme.statusEffects.hover)) {
    background = base[theme.statusEffects.hover.effect](
      theme.statusEffects.hover.factor,
    ).toRgbString();
  } else {
    background = base
      .mix(theme.statusEffects.hover.second, theme.statusEffects.hover.factor)
      .toRgbString();
  }

  return effectColorStyle(disabled, swapped, background, 'hover', {
    boxShadow: disabled ? undefined : `${topShadow(0)}, ${bottomShadow(0)}`,
    ...(outline ? { backgroundColor: 'transparent' } : undefined),
  });
};
