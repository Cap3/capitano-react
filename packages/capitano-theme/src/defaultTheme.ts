import { CapitanoTheme } from "./index";

// tslint:disable:max-file-line-count
export const defaultLightTheme: CapitanoTheme = {
  spacing: {
    baseUnit: 8,
    borderRadiusFactor: {
      sharp: 0,
      small: 0.5,
      medium: 1,
    },
    formElementFactor: {
      large: 6,
      medium: 4,
      small: 3,
    },
  },
  colors: {
    layout: {
      M0: "#000000",
      M10: "#333333",
      M20: "#5a5a5a",
      M30: "#787878",
      M40: "#949494",
      M50: "#b2b2b2",
      M60: "#cecece",
      M70: "#e6e6e6",
      M80: "#f7f7f7",
      M90: "#fcfcfc",
      base: "#fefefe",
    },
    textOnBackground: "#2b2b2b",
    textOnBackgroundInverse: "#fafafa",
    primary: {
      "50": "#e0eef9",
      "100": "#b3d5f1",
      "200": "#80b9e8",
      "300": "#4d9ddf",
      "400": "#2688d8",
      "500": "#0073d1",
      "600": "#006bcc",
      "700": "#0060c6",
      "800": "#0056c0",
      "900": "#0043b5",
      A100: "#dee8ff",
      A200: "#abc4ff",
      A400: "#78a0ff",
      A700: "#5e8eff",
    },
    textOnPrimary: "#fafafa",
    secondary: {
      "50": "#f6fae0",
      "100": "#e8f2b3",
      "200": "#d9ea80",
      "300": "#cae14d",
      "400": "#beda26",
      "500": "#b3d400",
      "600": "#accf00",
      "700": "#a3c900",
      "800": "#9ac300",
      "900": "#8bb900",
      A100: "#f7ffe2",
      A200: "#e8ffaf",
      A400: "#d9ff7c",
      A700: "#d2ff63",
    },
    textOnSecondary: "#fafafa",
    success: {
      "50": "#e1f8e9",
      "100": "#b5eec9",
      "200": "#84e3a5",
      "300": "#52d881",
      "400": "#2dcf66",
      "500": "#08c74b",
      "600": "#07c144",
      "700": "#06ba3b",
      "800": "#04b333",
      "900": "#02a623",
      A100: "#d0ffd7",
      A200: "#9dffab",
      A400: "#6aff7f",
      A700: "#51ff69",
    },
    textOnSuccess: "#fafafa",
    warning: {
      "50": "#fff9ed",
      "100": "#fff1d1",
      "200": "#ffe8b3",
      "300": "#ffdf94",
      "400": "#ffd87d",
      "500": "#ffd166",
      "600": "#ffcc5e",
      "700": "#ffc653",
      "800": "#ffc049",
      "900": "#ffb538",
      A100: "#ffffff",
      A200: "#ffffff",
      A400: "#fff3e1",
      A700: "#ffe9c8",
    },
    danger: {
      "50": "#fde3e9",
      "100": "#fab8c8",
      "200": "#f789a3",
      "300": "#f45a7e",
      "400": "#f13663",
      "500": "#ef1347",
      "600": "#ed1140",
      "700": "#eb0e37",
      "800": "#e80b2f",
      "900": "#e40620",
      A100: "#ffffff",
      A200: "#ffd9dc",
      A400: "#ffa6ad",
      A700: "#ff8c96",
    },
    textOnDanger: "#fafafa",
    disabled: {
      M0: "#000000",
      M10: "#111415",
      M20: "#23272a",
      M30: "#353b40",
      M40: "#475056",
      M50: "#5a656d",
      M60: "#6e7c86",
      M70: "#86939b",
      M80: "#9fa8b0",
      M90: "#b2bac0",
      base: "#bbc2c7",
    },
    textOnDisabled: "#fafafa",
  },
  statusEffects: {
    hover: {
      effect: "brighten",
      factor: 15,
    },
    pressed: {
      effect: "shade",
      factor: 15,
    },
    focus: {
      effect: "tint",
      factor: 20,
    },
  },
  typography: {
    fontFamily: {
      base: "Open Sans, sans-serif",
      monospace:
        '"Dank Mono", "Fira Code", Consolas, "Liberation Mono", Menlo, monospace',
    },
    fontSize: {
      medium: 14,
      large: 16,
      small: 12,
      h1: 32,
      h2: 28,
      h3: 24,
      h4: 20,
      h5: 16,
      h6: 14,
    },
  },
};

export const defaultDarkTheme: CapitanoTheme = {
  spacing: {
    baseUnit: 8,
    borderRadiusFactor: {
      sharp: 0,
      small: 0.5,
      medium: 1,
    },
    formElementFactor: {
      large: 6,
      medium: 4,
      small: 3,
    },
  },
  colors: {
    layout: {
      M0: "#ffffff",
      M10: "#f6f6f6",
      M20: "#eeeeee",
      M30: "#dcdcdc",
      M40: "#c8c8c8",
      M50: "#b4b4b4",
      M60: "#9d9d9d",
      M70: "#858585",
      M80: "#686868",
      M90: "#424242",
      base: "#252525",
    },
    textOnBackground: "#dedede",
    textOnBackgroundInverse: "#404040",
    primary: {
      "50": "#e0eef9",
      "100": "#b3d5f1",
      "200": "#80b9e8",
      "300": "#4d9ddf",
      "400": "#2688d8",
      "500": "#0073d1",
      "600": "#006bcc",
      "700": "#0060c6",
      "800": "#0056c0",
      "900": "#0043b5",
      A100: "#dee8ff",
      A200: "#abc4ff",
      A400: "#78a0ff",
      A700: "#5e8eff",
    },
    textOnPrimary: "#ffffff",
    secondary: {
      "50": "#f6fae0",
      "100": "#e8f2b3",
      "200": "#d9ea80",
      "300": "#cae14d",
      "400": "#beda26",
      "500": "#b3d400",
      "600": "#accf00",
      "700": "#a3c900",
      "800": "#9ac300",
      "900": "#8bb900",
      A100: "#f7ffe2",
      A200: "#e8ffaf",
      A400: "#d9ff7c",
      A700: "#d2ff63",
    },
    textOnSecondary: "#ffffff",
    success: {
      "50": "#e1f8e9",
      "100": "#b5eec9",
      "200": "#84e3a5",
      "300": "#52d881",
      "400": "#2dcf66",
      "500": "#08c74b",
      "600": "#07c144",
      "700": "#06ba3b",
      "800": "#04b333",
      "900": "#02a623",
      A100: "#d0ffd7",
      A200: "#9dffab",
      A400: "#6aff7f",
      A700: "#51ff69",
    },
    textOnSuccess: "#ffffff",
    warning: {
      "50": "#fff9ed",
      "100": "#fff1d1",
      "200": "#ffe8b3",
      "300": "#ffdf94",
      "400": "#ffd87d",
      "500": "#ffd166",
      "600": "#ffcc5e",
      "700": "#ffc653",
      "800": "#ffc049",
      "900": "#ffb538",
      A100: "#ffffff",
      A200: "#ffffff",
      A400: "#fff3e1",
      A700: "#ffe9c8",
    },
    danger: {
      "50": "#fde3e9",
      "100": "#fab8c8",
      "200": "#f789a3",
      "300": "#f45a7e",
      "400": "#f13663",
      "500": "#ef1347",
      "600": "#ed1140",
      "700": "#eb0e37",
      "800": "#e80b2f",
      "900": "#e40620",
      A100: "#ffffff",
      A200: "#ffd9dc",
      A400: "#ffa6ad",
      A700: "#ff8c96",
    },
    textOnDanger: "#ffffff",
    disabled: {
      M0: "#000000",
      M10: "#111415",
      M20: "#23272a",
      M30: "#353b40",
      M40: "#475056",
      M50: "#5a656d",
      M60: "#6e7c86",
      M70: "#86939b",
      M80: "#9fa8b0",
      M90: "#b2bac0",
      base: "#bbc2c7",
    },
    textOnDisabled: "#ffffff",
  },
  statusEffects: {
    hover: {
      effect: "brighten",
      factor: 15,
    },
    pressed: {
      effect: "shade",
      factor: 15,
    },
    focus: {
      effect: "tint",
      factor: 20,
    },
  },
  typography: {
    fontFamily: {
      base: "Open Sans, sans-serif",
      monospace:
        '"Dank Mono", "Fira Code", Consolas, "Liberation Mono", Menlo, monospace',
    },
    fontSize: {
      medium: 14,
      large: 16,
      small: 12,
      h1: 32,
      h2: 28,
      h3: 24,
      h4: 20,
      h5: 16,
      h6: 14,
    },
  },
};
