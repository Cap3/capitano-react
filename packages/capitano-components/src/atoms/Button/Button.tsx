import { styled, CapitanoTheme, spacingAbsolute } from "@cap3/capitano-theme";
import { focusEffect, hoverEffect, activeEffect } from "./mouseEffects";
import assertNever from "../../utils/assertNever";
import {
  FormElementRadius,
  FormElementSize,
  getFormElementBorderRadius,
} from "../basics/formElementHelpers";

export enum ButtonType {
  primary = "primary",
  secondary = "secondary",
  success = "success",
  danger = "danger",
}

type Props = {
  fixed?: boolean;
  type?: ButtonType;
  radius?: FormElementRadius;
  size?: FormElementSize;
  square?: boolean;
  outline?: boolean;
  disabled?: boolean;
};

const getColor = (
  type: ButtonType = ButtonType.primary,
  theme: CapitanoTheme,
): string => {
  switch (type) {
    case ButtonType.primary:
      return theme.colors.primary["500"];
    case ButtonType.secondary:
      return theme.colors.secondary["500"];
    case ButtonType.success:
      return theme.colors.success["500"];
    case ButtonType.danger:
      return theme.colors.danger["500"];
    default:
      return assertNever(type);
  }
};

const getTextColor = (
  type: ButtonType = ButtonType.primary,
  theme: CapitanoTheme,
): string => {
  switch (type) {
    case ButtonType.primary:
      return theme.colors.textOnPrimary;
    case ButtonType.secondary:
      return theme.colors.textOnSecondary;
    case ButtonType.success:
      return theme.colors.textOnSuccess;
    case ButtonType.danger:
      return theme.colors.textOnDanger;
    default:
      return assertNever(type);
  }
};

export const Button = styled("button")<Props>(
  {
    display: "flex",
    position: "relative",
    justifyContent: "center",
    alignItems: "center",
    border: "none",
    outline: "none",
    overflow: "hidden",
    padding: `0 ${spacingAbsolute.spx8}px`,
  },
  // colors for different types and states
  ({ theme, type, disabled = false, outline = false }) => {
    const backgroundColor = disabled
      ? theme.colors.disabled.base
      : getColor(type, theme);

    const color = disabled
      ? theme.colors.textOnDisabled
      : getTextColor(type, theme);
    const effectArgs = {
      disabled,
      swapped: outline,
      outline,
      baseColor: backgroundColor,
      theme,
    };
    const focus = focusEffect(effectArgs);
    const hover = hoverEffect(effectArgs);
    const active = activeEffect(effectArgs);

    if (outline) {
      return {
        borderWidth: "2px",
        borderColor: backgroundColor,
        borderStyle: "solid",
        backgroundColor: "transparent",
        color: backgroundColor,
        ...focus,
        ...hover,
        ...active,
      };
    } else {
      return {
        backgroundColor,
        color,
        ...focus,
        ...hover,
        ...active,
      };
    }
  },
  // text styles
  ({ theme }) => ({
    fontSize: theme.typography.fontSize.large,
    fontFamily: theme.typography.fontFamily.base,
  }),
  // cursor style
  ({ disabled }) => ({
    cursor: disabled ? "default" : "pointer",
  }),
  // button dimensions
  ({ radius, size, fixed, theme, square }) => {
    const height =
      theme.spacing.formElementFactor[size || "medium"] *
      theme.spacing.baseUnit;
    const borderRadius = getFormElementBorderRadius(radius, height, theme);

    const width = square ? height : fixed ? 120 : "auto";
    return { borderRadius, height, width };
  },
);
