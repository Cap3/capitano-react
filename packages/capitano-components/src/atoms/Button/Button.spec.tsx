import * as React from "react";
import { ThemeProvider } from "emotion-theming";
import { Button } from "./Button";
import { render } from "react-testing-library";
import { defaultLightTheme } from "@cap3/capitano-theme/lib/defaultTheme";

describe("DummyComponent", () => {
  test("render dummy snapshot with react-testing-library", () => {
    const wrapper = render(
      <ThemeProvider theme={defaultLightTheme}>
        <Button>basic button</Button>
      </ThemeProvider>,
    );
    expect(wrapper.container.firstChild).toMatchSnapshot();
  });
});
