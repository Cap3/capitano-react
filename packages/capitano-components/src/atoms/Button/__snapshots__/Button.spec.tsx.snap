// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DummyComponent render dummy snapshot with react-testing-library 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  outline: none;
  overflow: hidden;
  padding: 0 8px;
  background-color: #2b597e;
  color: #ffffff;
  font-size: 16px;
  font-family: Open Sans,sans-serif;
  cursor: pointer;
  border-radius: 8px;
  height: 32px;
  width: auto;
}

.emotion-0:focus {
  background-color: rgb(85,122,152);
}

.emotion-0:hover {
  background-color: rgb(81,127,164);
  box-shadow: 0 1.5px 6px rgba(0,0,0,0.12),0 1.5px 6px rgba(0,0,0,0.24);
}

.emotion-0:active {
  background-color: rgb(37,76,107);
}

<button
  class="emotion-0"
>
  basic button
</button>
`;
